Classes/Service/QbankService.php

```diff
@@ -189,6 +219,221 @@ public function removeMediaUsageInFileReference(int $fileReferenceId): void
+            $targetProperty = $metaDataMappings[$sourceProperty];
+
+            if (!isset($targetProperty)) {
+                continue;
+            }
+
+            $this->eventDispatcher->dispatch(new FilePropertyChangeEvent(
+                $file,
+                $targetProperty,
+                PropertyUtility::convertQbankToFileProperty(
+                    $qbankPropertyValue,
+                    $targetProperty
+                )
+            ));
+        }
+
+        $this->updateFileRecord($file->getUid(), false, true, $qbankId);
+
+        $this->eventDispatcher->dispatch(new AfterFilePropertyChangesEvent($file));
+    }
+
+    /**
+     * Synchronize file content for a particular file UID.
+     *
+     * @param int $fileId The FAL file UID
+     * @throws \TYPO3\CMS\Core\Resource\Exception\FileDoesNotExistException
+     */
+    public function replaceLocalMedia(int $fileId): void
+    {
+        $file = $this->resourceFactory->getFileObject($fileId);
+        if ($file === null) {
+            throw new \TYPO3\CMS\Core\Resource\Exception\FileDoesNotExistException(
+                'No file found for given UID: ' . $fileId,
+                1623070299
+            );
+        }
+
+        $replacedByMediaIdentifier = $this->getQbankReplacedByMediaIdentifierForFile($fileId);
+
+        $replacedByFile = $this->createLocalMediaCopy($replacedByMediaIdentifier);
+        if ($replacedByFile === null) {
+            throw new \TYPO3\CMS\Core\Resource\Exception\FileDoesNotExistException(
+                'No file found for given QBank id: ' . $replacedByMediaIdentifier,
+                1623306399
+            );
+        }
+
+        $sysFileReferences = GeneralUtility::makeInstance(SysFileReferenceRepository::class)
+            ->fetchRawSysFileReferencesByFileId($fileId);
+
+        $data = [];
+        $cmd = [];
+        $pids = [];
+
+        // Go through references to build data for sys_file_reference updates.
+        if (is_array($sysFileReferences)) {
+            foreach ($sysFileReferences as $sysFileReference) {
+                $this->removeMediaUsageInFileReference($sysFileReference['uid']);
+
+                $newId = StringUtility::getUniqueId('NEW');
+
+                $data['sys_file_reference'][$newId] = $sysFileReference;
+
+                unset($data['sys_file_reference'][$newId]['uid']);
+
+                $data['sys_file_reference'][$newId]['uid_local'] = $replacedByFile->getProperty('uid');
+
+                $pids[] = $sysFileReference['pid'];
+
+                $cmd['sys_file_reference'][$sysFileReference['uid']]['delete'] = 1;
+            }
+
+            if (count($data) > 0) {
+                /** @var DataHandler $dataHandler */
+                $dataHandler = GeneralUtility::makeInstance(DataHandler::class);
+                $dataHandler->start($data, $cmd);
+                $dataHandler->process_datamap();
+                $dataHandler->process_cmdmap();
```

Comment:
Missing error handling. Should throw an exception if the DataHandler has errors.